File: src/researchclaw/research/smart_sectioning.py
# Copyright 2026 OpenClaw
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Smart Sectioning - Automatic chapter detection and theme recognition
"""

from typing import List, Dict, Any, Optional, Tuple
from dataclasses import dataclass, field
import logging

logger = logging.getLogger(__name__)


@dataclass
class SectionCandidate:
    """A potential section detected in content"""
    title: str
    content: str
    start_position: int = 0
    end_position: int = 0
    confidence: float = 0.0
    theme: str = ""
    keywords: List[str] = field(default_factory=list)
    parent_section: Optional[str] = None


@dataclass
class SectionAnalysis:
    """Analysis result for smart sectioning"""
    sections: List[SectionCandidate]
    themes: List[str] = field(default_factory=list)
    structure_score: float = 0.0
    suggested_order: List[str] = field(default_factory=list)


class SmartSectioner:
    """Automatic section detection and theme recognition"""

    # Common research section patterns
    SECTION_PATTERNS = {
        "introduction": ["introduction", "background", "overview", "context"],
        "methods": ["method", "approach", "methodology", "research design", "procedure"],
        "results": ["result", "finding", "observation", "data", "analysis"],
        "discussion": ["discussion", "interpretation", "implication", "limitations"],
        "conclusion": ["conclusion", "summary", "takeaway", "final"],
        "references": ["reference", "bibliography", "citation", "source"],
    }

    def __init__(self, llm_engine=None):
        """Initialize smart sectioner
        
        Args:
            llm_engine: LLM engine for advanced analysis
        """
        self.llm_engine = llm_engine

    def detect_sections(
        self,
        content: str,
        min_section_length: int = 100
    ) -> SectionAnalysis:
        """Detect sections in content
        
        Args:
            content: Raw content text
            min_section_length: Minimum section length
            
        Returns:
            SectionAnalysis with detected sections
        """
        sections = []
        
        # Try LLM-based detection first
        if self.llm_engine:
            try:
                return self._detect_sections_llm(content, min_section_length)
            except Exception as e:
                logger.warning(f"LLM section detection failed: {e}")
        
        # Fallback to pattern-based detection
        return self._detect_sections_pattern(content, min_section_length)

    def _detect_sections_llm(
        self,
        content: str,
        min_section_length: int
    ) -> SectionAnalysis:
        """Detect sections using LLM
        
        Args:
            content: Raw content
            min_section_length: Minimum section length
            
        Returns:
            SectionAnalysis
        """
        import json
        
        # Split content if too long
        chunks = self._split_content(content, max_chars=4000)
        
        prompt = f"""Analyze the following text and identify the main sections/subsections.

Text:
{content[:3000]}

Respond with a JSON array of sections, each with:
- "title": section title
- "content": main content of this section (abbreviated if long)
- "confidence": how confident you are this is a real section (0-1)
- "keywords": key terms in this section
- "theme": main theme/category

Return ONLY valid JSON."""

        response = self.llm_engine.chat_simple(
            prompt=prompt,
            system_prompt="You are a document structure analyzer. Always respond with valid JSON."
        )
        
        sections_data = json.loads(response)
        
        sections = [
            SectionCandidate(
                title=s.get("title", "Untitled"),
                content=s.get("content", ""),
                confidence=s.get("confidence", 0.5),
                keywords=s.get("keywords", []),
                theme=s.get("theme", ""),
            )
            for s in sections_data
        ]
        
        # Extract unique themes
        themes = list(set(s.theme for s in sections if s.theme))
        
        return SectionAnalysis(
            sections=sections,
            themes=themes,
            structure_score=sum(s.confidence for s in sections) / max(len(sections), 1),
            suggested_order=[s.title for s in sections],
        )

    def _detect_sections_pattern(
        self,
        content: str,
        min_section_length: int
    ) -> SectionAnalysis:
        """Detect sections using pattern matching
        
        Args:
            content: Raw content
            min_section_length: Minimum section length
            
        Returns:
            SectionAnalysis
        """
        import re
        
        sections = []
        lines = content.split("\n")
        
        current_section = None
        current_content = []
        current_pos = 0
        
        # Header patterns
        header_pattern = re.compile(r"^(#{1,6})\s+(.+)$|^([A-Z][^\.]{5,50})\.\s*$", re.MULTILINE)
        
        for i, line in enumerate(lines):
            # Check for markdown headers
            header_match = header_pattern.match(line)
            
            if header_match:
                # Save previous section
                if current_section and current_content:
                    content_text = "\n".join(current_content)
                    if len(content_text) >= min_section_length:
                        current_section.content = content_text
                        current_section.end_position = current_pos
                        sections.append(current_section)
                
                # Start new section
                title = header_match.group(2) or header_match.group(3)
                current_section = SectionCandidate(
                    title=title.strip(),
                    content="",
                    start_position=current_pos,
                    confidence=0.7,
                )
                current_content = []
            
            current_pos += len(line) + 1
            if current_section:
                current_content.append(line)
        
        # Save last section
        if current_section and current_content:
            current_section.content = "\n".join(current_content)
            current_section.end_position = current_pos
            sections.append(current_section)
        
        # If no sections detected, create single section
        if not sections:
            sections.append(SectionCandidate(
                title="Overview",
                content=content,
                confidence=0.5,
            ))
        
        # Identify themes
        themes = self._identify_themes(sections)
        
        return SectionAnalysis(
            sections=sections,
            themes=themes,
            structure_score=0.6,
            suggested_order=[s.title for s in sections],
        )

    def _identify_themes(self, sections: List[SectionCandidate]) -> List[str]:
        """Identify themes from sections
        
        Args:
            sections: List of sections
            
        Returns:
            List of theme names
        """
        theme_counts: Dict[str, int] = {}
        
        for section in sections:
            # Check against known patterns
            title_lower = section.title.lower()
            
            for theme, keywords in self.SECTION_PATTERNS.items():
                if any(kw in title_lower for kw in keywords):
                    theme_counts[theme] = theme_counts.get(theme, 0) + 1
        
        # Return themes sorted by frequency
        sorted_themes = sorted(theme_counts.items(), key=lambda x: -x[1])
        return [t[0] for t in sorted_themes]

    def group_by_theme(
        self,
        sections: List[SectionCandidate]
    ) -> Dict[str, List[SectionCandidate]]:
        """Group sections by theme
        
        Args:
            sections: List of sections
            
        Returns:
            Dict mapping theme to sections
        """
        theme_groups: Dict[str, List[SectionCandidate]] = {}
        
        for section in sections:
            # Determine theme
            theme = section.theme or self._classify_section_theme(section)
            
            if theme not in theme_groups:
                theme_groups[theme] = []
            theme_groups[theme].append(section)
        
        return theme_groups

    def _classify_section_theme(self, section: SectionCandidate) -> str:
        """Classify section into a theme
        
        Args:
            section: Section to classify
            
        Returns:
            Theme name
        """
        title_lower = section.title.lower()
        content_lower = section.content.lower()[:500]
        
        # Check patterns
        for theme, keywords in self.SECTION_PATTERNS.items():
            if any(kw in title_lower for kw in keywords):
                return theme
            if any(kw in content_lower for kw in keywords):
                return theme
        
        return "general"

    def optimize_order(
        self,
        sections: List[SectionCandidate]
    ) -> List[SectionCandidate]:
        """Optimize section ordering
        
        Args:
            sections: Sections to order
            
        Returns:
            Ordered list of sections
        """
        # Standard research paper order
        preferred_order = [
            "introduction", "background", "overview",
            "methods", "methodology", "approach",
            "results", "findings", "analysis",
            "discussion", "interpretation",
            "conclusion", "summary",
        ]
        
        def get_order_score(section: SectionCandidate) -> int:
            theme = self._classify_section_theme(section)
            try:
                return preferred_order.index(theme)
            except ValueError:
                return len(preferred_order)
        
        return sorted(sections, key=get_order_score)

    def _split_content(self, content: str, max_chars: int = 4000) -> List[str]:
        """Split content into chunks
        
        Args:
            content: Content to split
            max_chars: Max chars per chunk
            
        Returns:
            List of content chunks
        """
        if len(content) <= max_chars:
            return [content]
        
        chunks = []
        paragraphs = content.split("\n\n")
        current_chunk = ""
        
        for para in paragraphs:
            if len(current_chunk) + len(para) > max_chars:
                if current_chunk:
                    chunks.append(current_chunk)
                current_chunk = para
            else:
                current_chunk += "\n\n" + para
        
        if current_chunk:
            chunks.append(current_chunk)
        
        return chunks[:10]  # Limit chunks

    def calculate_accuracy(
        self,
        detected: List[SectionCandidate],
        ground_truth: List[str]
    ) -> float:
        """Calculate section detection accuracy
        
        Args:
            detected: Detected section titles
            ground_truth: Expected section titles
            
        Returns:
            Accuracy score (0-1)
        """
        if not ground_truth:
            return 0.0
        
        # Normalize titles for comparison
        def normalize(title: str) -> str:
            return title.lower().strip()
        
        detected_normalized = {normalize(t) for t in detected}
        truth_normalized = {normalize(t) for t in ground_truth}
        
        # Calculate precision and recall
        true_positives = len(detected_normalized & truth_normalized)
        precision = true_positives / len(detected_normalized) if detected_normalized else 0
        recall = true_positives / len(truth_normalized) if truth_normalized else 0
        
        # F1 score
        if precision + recall == 0:
            return 0.0
        
        f1 = 2 * (precision * recall) / (precision + recall)
        return f1


__all__ = [
    "SectionCandidate",
    "SectionAnalysis", 
    "SmartSectioner",
]

File: src/research/smart_sectioning.ts
/**
 * Smart Sectioning - Automatic chapter detection and theme recognition
 *
 * Copyright 2026 OpenClaw
 * Licensed under the Apache License, Version 2.0
 */

import type { SectionCandidate, SectionAnalysis, LLMEngine } from "./types.js";

/**
 * Common research section patterns
 */
const SECTION_PATTERNS: Record<string, string[]> = {
  introduction: ["introduction", "background", "overview", "context"],
  methods: ["method", "approach", "methodology", "research design", "procedure"],
  results: ["result", "finding", "observation", "data", "analysis"],
  discussion: ["discussion", "interpretation", "implication", "limitations"],
  conclusion: ["conclusion", "summary", "takeaway", "final"],
  references: ["reference", "bibliography", "citation", "source"],
};

/**
 * Automatic section detection and theme recognition
 */
export class SmartSectioner {
  private llmEngine: LLMEngine | null;

  constructor(llmEngine?: LLMEngine) {
    this.llmEngine = llmEngine ?? null;
  }

  /**
   * Detect sections in content
   */
  async detectSections(
    content: string,
    minSectionLength: number = 100
  ): Promise<SectionAnalysis> {
    // Try LLM-based detection first
    if (this.llmEngine) {
      try {
        return await this.detectSectionsLLM(content, minSectionLength);
      } catch {
        // Fall through to pattern-based detection
      }
    }

    // Fallback to pattern-based detection
    return this.detectSectionsPattern(content, minSectionLength);
  }

  /**
   * Detect sections using LLM
   */
  private async detectSectionsLLM(
    content: string,
    _minSectionLength: number
  ): Promise<SectionAnalysis> {
    const prompt = `Analyze the following text and identify the main sections/subsections.

Text:
${content.slice(0, 3000)}

Respond with a JSON array of sections, each with:
- "title": section title
- "content": main content of this section (abbreviated if long)
- "confidence": how confident you are this is a real section (0-1)
- "keywords": key terms in this section
- "theme": main theme/category

Return ONLY valid JSON.`;

    const response = await this.llmEngine!.chatSimple(
      prompt,
      "You are a document structure analyzer. Always respond with valid JSON."
    );

    const sectionsData = JSON.parse(response) as Array<{
      title: string;
      content: string;
      confidence: number;
      keywords: string[];
      theme: string;
    }>;

    const sections: SectionCandidate[] = sectionsData.map((s) => ({
      title: s.title || "Untitled",
      content: s.content || "",
      startPosition: 0,
      endPosition: 0,
      confidence: s.confidence ?? 0.5,
      keywords: s.keywords || [],
      theme: s.theme || "",
    }));

    // Extract unique themes
    const themes = [...new Set(sections.map((s) => s.theme).filter(Boolean))];

    return {
      sections,
      themes,
      structureScore:
        sections.reduce((sum, s) => sum + s.confidence, 0) / Math.max(sections.length, 1),
      suggestedOrder: sections.map((s) => s.title),
    };
  }

  /**
   * Detect sections using pattern matching
   */
  private detectSectionsPattern(
    content: string,
    minSectionLength: number
  ): SectionAnalysis {
    const sections: SectionCandidate[] = [];
    const lines = content.split("\n");

    let currentSection: SectionCandidate | null = null;
    const currentContent: string[] = [];
    let currentPos = 0;

    // Header patterns
    const headerPattern = /^(#{1,6})\s+(.+)$|^([A-Z][^.]{5,50})\.\s*$/;

    for (const line of lines) {
      // Check for markdown headers
      const headerMatch = line.match(headerPattern);

      if (headerMatch) {
        // Save previous section
        if (currentSection && currentContent.length > 0) {
          const contentText = currentContent.join("\n");
          if (contentText.length >= minSectionLength) {
            currentSection.content = contentText;
            currentSection.endPosition = currentPos;
            sections.push(currentSection);
          }
        }

        // Start new section
        const title = headerMatch[2] || headerMatch[3];
        if (!title) continue;
        currentSection = {
          title: title.trim(),
          content: "",
          startPosition: currentPos,
          endPosition: 0,
          confidence: 0.7,
          theme: "",
          keywords: [],
        };
        currentContent.length = 0;
      }

      currentPos += line.length + 1;
      if (currentSection) {
        currentContent.push(line);
      }
    }

    // Save last section
    if (currentSection && currentContent.length > 0) {
      currentSection.content = currentContent.join("\n");
      currentSection.endPosition = currentPos;
      sections.push(currentSection);
    }

    // If no sections detected, create single section
    if (sections.length === 0) {
      sections.push({
        title: "Overview",
        content: content,
        startPosition: 0,
        endPosition: content.length,
        confidence: 0.5,
        theme: "",
        keywords: [],
      });
    }

    // Identify themes
    const themes = this.identifyThemes(sections);

    return {
      sections,
      themes,
      structureScore: 0.6,
      suggestedOrder: sections.map((s) => s.title),
    };
  }

  /**
   * Identify themes from sections
   */
  private identifyThemes(sections: SectionCandidate[]): string[] {
    const themeCounts = new Map<string, number>();

    for (const section of sections) {
      const titleLower = section.title.toLowerCase();

      for (const [theme, keywords] of Object.entries(SECTION_PATTERNS)) {
        if (keywords.some((kw) => titleLower.includes(kw))) {
          themeCounts.set(theme, (themeCounts.get(theme) || 0) + 1);
        }
      }
    }

    // Return themes sorted by frequency
    return [...themeCounts.entries()]
      .sort((a, b) => b[1] - a[1])
      .map(([theme]) => theme);
  }

  /**
   * Group sections by theme
   */
  groupByTheme(sections: SectionCandidate[]): Map<string, SectionCandidate[]> {
    const themeGroups = new Map<string, SectionCandidate[]>();

    for (const section of sections) {
      const theme = section.theme || this.classifySectionTheme(section);

      if (!themeGroups.has(theme)) {
        themeGroups.set(theme, []);
      }
      themeGroups.get(theme)!.push(section);
    }

    return themeGroups;
  }

  /**
   * Classify section into a theme
   */
  private classifySectionTheme(section: SectionCandidate): string {
    const titleLower = section.title.toLowerCase();
    const contentLower = section.content.toLowerCase().slice(0, 500);

    for (const [theme, keywords] of Object.entries(SECTION_PATTERNS)) {
      if (keywords.some((kw) => titleLower.includes(kw))) {
        return theme;
      }
      if (keywords.some((kw) => contentLower.includes(kw))) {
        return theme;
      }
    }

    return "general";
  }

  /**
   * Optimize section ordering
   */
  optimizeOrder(sections: SectionCandidate[]): SectionCandidate[] {
    const preferredOrder = [
      "introduction",
      "background",
      "overview",
      "methods",
      "methodology",
      "approach",
      "results",
      "findings",
      "analysis",
      "discussion",
      "interpretation",
      "conclusion",
      "summary",
    ];

    const getOrderScore = (section: SectionCandidate): number => {
      const theme = this.classifySectionTheme(section);
      const index = preferredOrder.indexOf(theme);
      return index >= 0 ? index : preferredOrder.length;
    };

    return [...sections].sort((a, b) => getOrderScore(a) - getOrderScore(b));
  }

  /**
   * Split content into chunks
   */
  private splitContent(content: string, maxChars: number = 4000): string[] {
    if (content.length <= maxChars) {
      return [content];
    }

    const chunks: string[] = [];
    const paragraphs = content.split("\n\n");
    let currentChunk = "";

    for (const para of paragraphs) {
      if (currentChunk.length + para.length > maxChars) {
        if (currentChunk) {
          chunks.push(currentChunk);
        }
        currentChunk = para;
      } else {
        currentChunk += "\n\n" + para;
      }
    }

    if (currentChunk) {
      chunks.push(currentChunk);
    }

    return chunks.slice(0, 10); // Limit chunks
  }

  /**
   * Calculate section detection accuracy
   */
  calculateAccuracy(
    detected: SectionCandidate[],
    groundTruth: string[]
  ): number {
    if (groundTruth.length === 0) {
      return 0.0;
    }

    // Normalize titles for comparison
    const normalize = (title: string): string =>
      title.toLowerCase().trim();

    const detectedNormalized = new Set(detected.map((s) => normalize(s.title)));
    const truthNormalized = new Set(groundTruth.map(normalize));

    // Calculate precision and recall
    let truePositives = 0;
    for (const title of detectedNormalized) {
      if (truthNormalized.has(title)) {
        truePositives++;
      }
    }

    const precision =
      detectedNormalized.size > 0
        ? truePositives / detectedNormalized.size
        : 0;
    const recall =
      truthNormalized.size > 0 ? truePositives / truthNormalized.size : 0;

    // F1 score
    if (precision + recall === 0) {
      return 0.0;
    }

    return (2 * (precision * recall)) / (precision + recall);
  }
}

/**
 * Create default sectioner instance
 */
export const smartSectioner = new SmartSectioner();